File: examples/cpp/src/skeleton.cpp
#include <qnbot/glove.hpp>

#include "example_support.hpp"

#include <cstdlib>
#include <iostream>
#include <optional>
#include <string>

int main(int argc, char** argv) {
    try {
        std::string port;
        std::optional<qnbot::Side> side;
        for (int index = 1; index < argc; ++index) {
            const std::string argument = argv[index];
            if (argument == "--port") {
                port = example::require_value(argc, argv, index, argument);
            } else if (argument == "--side") {
                side = example::parse_side(
                    example::require_value(argc, argv, index, argument));
            } else {
                throw std::invalid_argument("unknown argument: " + argument);
            }
        }
        if (port.empty()) throw std::invalid_argument("--port is required");
        if (!side) throw std::invalid_argument("--side is required");

        qnbot::Sdk sdk(example::serial_config(port, *side));
        auto glove = sdk.glove();

        auto skeleton = glove.device().skeleton();
        const auto joint_angles_subscription =
            skeleton.joint_angles().subscribe(
                [](const qnbot::Sample<qnbot::HandJointCommand>& sample) {
                    std::cout << "skeleton sequence=" << sample.sequence
                              << " target=" << sample.value.target
                              << " joints=" << sample.value.joints.size()
                              << '\n';
                });
        const auto pose_subscription = skeleton.pose().subscribe(
            [](const qnbot::Sample<qnbot::HandSkeletonPose>& sample) {
                const auto& pose = sample.value;
                std::cout << "pose sequence=" << sample.sequence
                          << " frame=" << pose.coordinate_frame << " wrist=["
                          << pose.positions_local[0][0] << ", "
                          << pose.positions_local[0][1] << ", "
                          << pose.positions_local[0][2] << "] index_tip=["
                          << pose.positions_local[9][0] << ", "
                          << pose.positions_local[9][1] << ", "
                          << pose.positions_local[9][2] << "]\n";
            });

        glove.start();
        std::cout << "running; press Ctrl+C to stop" << std::endl;
        glove.run_forever();
        static_cast<void>(joint_angles_subscription);
        static_cast<void>(pose_subscription);
        return EXIT_SUCCESS;
    } catch (const std::exception& error) {
        return example::report_error(error);
    }
}

File: examples/python/src/skeleton.py
from __future__ import annotations

import argparse
from threading import Lock

from qnbot_sdk import Sample, Sdk, SerialConnection, Side
from qnbot_sdk.glove import GloveConfig, HandJointCommand, HandSkeletonPose

_PRINT_LOCK = Lock()


def print_joint_angles(sample: Sample[HandJointCommand]) -> None:
    with _PRINT_LOCK:
        print(
            f"skeleton sequence={sample.sequence} "
            f"target={sample.value.target} joints={sample.value.joints}"
        )


def print_pose(sample: Sample[HandSkeletonPose]) -> None:
    wrist = sample.value.positions_local[0]
    index_tip = sample.value.positions_local[9]
    with _PRINT_LOCK:
        print(
            f"pose sequence={sample.sequence} frame={sample.value.coordinate_frame} "
            f"wrist={wrist} index_tip={index_tip}"
        )


def main() -> None:
    arguments = argparse.ArgumentParser(description="Read SDK-owned skeleton output")
    arguments.add_argument("--port", required=True, help="Serial port for the glove")
    arguments.add_argument(
        "--side",
        choices=(Side.LEFT.value, Side.RIGHT.value),
        required=True,
        help="Physical glove side",
    )
    options = arguments.parse_args()
    sdk = Sdk(
        devices=(
            GloveConfig(
                side=Side(options.side),
                connection=SerialConnection(port=options.port),
            ),
        )
    )
    glove = sdk.glove()
    try:
        glove.connect()
        device = glove.device()
        skeleton = device.skeleton()
        angle_subscription = skeleton.joint_angles().subscribe(print_joint_angles)
        pose_subscription = skeleton.pose().subscribe(print_pose)
        glove.start()

        try:
            glove.run_forever()
        except KeyboardInterrupt:
            print("\nstopping")
        finally:
            del pose_subscription
            del angle_subscription
    finally:
        glove.close()


if __name__ == "__main__":
    main()
